packages/openactive-broker-microservice/src/validator/validator-inputs.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Items are sent to the Validator Worker Pool (VWP) by saving them to files, which the VWP then reads and processes.
+ *
+ * The reason that they are sent as files, rather than, say maintained in memory, is so that there is a hard limit on
+ * memory usage from queuing items to Validator.
+ *
+ * The potential worst case scenario is one in which Validator takes a long time to start the first time e.g. because
+ * the OpenActive JSON-LD context is taking a long time to load.
+ * If the entire feed is harvested before validation has begun, this means that the entire feed will be held in memory
+ * waiting for validation.
+ */
+const fs = require('fs').promises;
+const fsExtra = require('fs-extra');
+const itertools = require('iter-tools');
+const mkdirp = require('mkdirp');
+const path = require('path');
+const { ITEM_VALIDATION_MODE, VALIDATOR_INPUT_TMP_DIR } = require('../broker-config');
+
+/**
+ * A chunk length of 100 gives a memory usage by Validator Workers of roughly up to 4kb * 100 * <NUM CORES>.
+ *
+ * - A rough average size of medium-size (e.g. SessionSeries rather than ScheduledSession) OpenActive items
+ *   (uncompressed) is about 2-4kb.
+ * - So, with 8 cores, the memory usage will be roughly up to 4kb * 100 * 8 = 3.2MB, which is acceptable.
+ * - To clarify, this is the memory usage of just holding this data in memory. This is not counting the additional
+ *   memory usage required by validator's mechanisms.
+ *
+ * The flipside is that this results in a number of validator-input files created that is roughly:
+ *
+ * - <NUM UPDATED ITEMS IN FEEDS> / 100
+ * - This would happen if validator took so long to run that all the input files accumulated before it had run even
+ *   once.
+ * - So, with 1M items (as a very high example!) in the feeds, this could result in 10,000 files. This also seems
+ *   acceptably unlikely to cause issues (e.g. inode usage issues).
+ */
+const VALIDATOR_ITEMS_CHUNK_LENGTH = 100;
+
+let validatorInputFilenameSequenceNum = 0;
+
+/**
+ * Call this before saving any Validator inputs.
+ */
+async function setUpValidatorInputs() {
+  await mkdirp(VALIDATOR_INPUT_TMP_DIR);
+  // Clear this directory of any files which may not have been cleaned up in a prior run
+  await clearValidatorInputTmpDir();
+}
+
+/**
+ * Call this after Validator Worker Pool has stopped.
+ */
+async function cleanUpValidatorInputs() {
+  // Once validator has stopped, we may as well clean this directory to free up storage space
+  await clearValidatorInputTmpDir();
+}
+
+/**
+ * For a given page of items, split out those which are validatable (i.e. ignore deleted ones) into files and save
```

Comment:
Are deleted items not validated? I vaguely remember seeing deleted items with a `data` object which is obviously wrong and should be flagged? Not even sure the validator validates against that usecase tbh

Reply:
Yes. Deleted items are not validated. This is the same as the previous logic. I agree it would be better to validate more wholistically but maybe one for another PR? Only reason I can think that it wasn't done before is maybe performance but probably just an oversight

Reply:
Ok sure 👍🏽